phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexUpgradeTool.java

```diff
@@ -0,0 +1,537 @@
+        dryRun = (cmdLine.hasOption(DRY_RUN_OPTION.getOpt()));
+    }
+
+    @VisibleForTesting
+    public void prepareToolSetup() {
+        try {
+            if (logFile != null) {
+                FileHandler fh = new FileHandler(logFile);
+                fh.setFormatter(new SimpleFormatter());
+                LOGGER.addHandler(fh);
+            }
+
+            prop.put(Indexer.INDEX_BUILDER_CONF_KEY, PhoenixIndexBuilder.class.getName());
+            prop.put(NonTxIndexBuilder.CODEC_CLASS_NAME_KEY, PhoenixIndexCodec.class.getName());
+
+            if (inputTables == null) {
+                inputTables = new String(Files.readAllBytes(Paths.get(inputFile)));
+            }
+            if (inputTables == null) {
+                LOGGER.severe("tables' list is not available; use -tb or -f option");
+                System.exit(1);
+            }
+            LOGGER.info("list of tables/indexes passed: " + inputTables);
+
+            tablesAndIndexes = extractTablesAndIndexes(inputTables);
+
+            if (operation.equalsIgnoreCase(UPGRADE_OP)) {
+                upgrade = true;
+            } else if (operation.equalsIgnoreCase(ROLLBACK_OP)) {
+                upgrade = false;
+            } else {
+                throw new IllegalStateException("Invalid option provided for "
+                        + OPERATION_OPTION.getOpt() + " expected values: {upgrade, rollback}");
+            }
+            if (dryRun) {
+                LOGGER.info("This is the beginning of the tool with dry run.");
+            }
+        } catch (IOException e) {
+            LOGGER.severe("Something went wrong "+e);
+        }
+    }
+
+    @VisibleForTesting
+    public void executeTool() {
+        try (Connection conn = DriverManager.getConnection(connectionURL, new Properties())) {
+            ConnectionQueryServices queryServices = conn.unwrap(PhoenixConnection.class)
+                    .getQueryServices();
+            Configuration conf = queryServices.getConfiguration();
+            executeTool(conn, queryServices, conf);
+        } catch (SQLException e) {
+            LOGGER.severe("Something went wrong in executing tool "+ e);
+        }
+    }
+
+    private void executeTool(Connection conn, ConnectionQueryServices queryServices,
+            Configuration conf) {
+
+        LOGGER.info("Executing "+ operation);
+
+        for (Map.Entry<String, ArrayList<String>> entry: tablesAndIndexes.entrySet()) {
+            String dataTableFullName = entry.getKey();
+            ArrayList<String> indexes = entry.getValue();
+
+            try (Admin admin = queryServices.getAdmin()) {
+
+                PTable dataTable = PhoenixRuntime.getTableNoCache(conn, dataTableFullName);
+                LOGGER.fine("Executing "+operation+" for " + dataTableFullName);
+
+                boolean mutable = !(dataTable.isImmutableRows());
+
+                if (upgrade) {
+                    setCompactionEnabled(admin, entry, false);
```

Comment:
Nit: Instead of passing-around entry to all these helper functions, they would be much more intuitive if you passed them (String tableName, ArrayList<String> indexNames) explicitly - one of both of them, as needed by these functions.

Reply:
@swaroopak This was the comment that I mentioned yesterday.

Reply:
Aah! Noted. :)